contracts/registry/src/lib.rs

```diff
@@ -86,29 +86,105 @@ impl Contract {
         &mut self,
         to: AccountId,
         #[allow(unused_variables)] memo: Option<String>,
-    ) -> (AccountId, TokenId, bool) {
+    ) -> (u32, bool) {
+        // TODO: test what is the max safe amount of updates
+        self._sbt_soul_transfer(to, 200)
+    }
+
+    // execution of the sbt_soult_transfer in this function to parametrize `max_updates` in
+    // order to facilitate tests.
+    pub(crate) fn _sbt_soul_transfer(&mut self, to: AccountId, limit: usize) -> (u32, bool) {
         let owner = env::predecessor_account_id();
-        let start = match self.ongoing_soul_tx.get(&to) {
+        let (resumed, start) = match self.ongoing_soul_tx.get(&to) {
             // starting the process
             None => {
+                require!(!self._is_banned(&to), "`to` is banned");
                 // insert into banlist and assuer owner is not already banned.
                 require!(
                     self.banlist.insert(&owner),
                     "caller banned: can't make soul transfer"
                 );
-                require!(!self._is_banned(&to), "`to` is banned");
-                emit_soul_transfer(&owner, &to);
-                IssuerTokenId {
-                    issuer_id: 0,
-                    token: 0,
-                }
+                (
+                    false,
+                    IssuerTokenId {
+                        issuer_id: 0,
+                        token: 0, // NOTE: this is class ID
+                    },
+                )
             }
             // resuming Soul Transfer process
-            Some(s) => s,
+            Some(s) => (true, s),
         };
 
-        println!("Starting at: {} {}", start.issuer_id, start.token);
-        env::panic_str("not implemented");
+        let batch: Vec<(BalanceKey, TokenId)> = self
+            .balances
+            .iter_from(BalanceKey {
+                owner: owner.clone(),
+                issuer_id: start.issuer_id,
+                class_id: start.token,
+            })
+            .take(limit)
+            .collect();
+
+        let mut b_new = BalanceKey {
+            owner: to.clone(),
+            issuer_id: 0,
+            class_id: 0,
+        };
+        let mut prev_issuer: IssuerId = 0;
+        let mut i = 0;
+        for (b, tid) in &batch {
```

Comment:
I think we could use `batch_key` and token_id` to make it easier to read and understand

Reply:
renamed to `key, token_id`